[코일] 로또 미션 제출합니다.

takhyeongmin opened the pull request:
안녕하세요!

바쁘신 와중에 제 코드를 읽어주셔서 감사합니다!

시간에 쫓겨 프로젝트 구조를 짜다보니 부족한게 너무 많이 보이는 것 같아요...!

개선되는 모습 보이겠습니다!

혹시 궁금한 게 있으면 리뷰어님께 DM으로 여쭤봐도 될까요 ?

kimhyochan committed:
[docs] README 작성

kimhyochan committed:
[feat] 구매 개수 계산 메서드 작성

생성자로는 구매 금액으로부터 구매 개수를 산출하는
로직을 표현할 수가 없다고 판단하여 정적 팩토리 메서드 사용했습니다.

kimhyochan committed:
[feat] 구매 금액 입력

kimhyochan committed:
[feat] 구매 개수 출력

kimhyochan committed:
[refactor] 구매 개수 클래스에 getter 추가

kimhyochan committed:
[feat] 로또 넘버 생성 추가 및 예외 처리

로또 넘버의 유효성 검증을 한다.

kimhyochan committed:
[feat] 로또 넘버 리스트 일급 컬렉션 추가

로또 넘버들을 자료구조로 로또 넘버들을 관리하고 값 도출하는 용도

kimhyochan committed:
[feat] 중복없는 랜덤값 생성하는 클래스 작성

kimhyochan committed:
[refactor] 로또 숫자 enum 으로 변경

로또 숫자 재생성 비용을 enum 으로 변경함으로써 단 한번의 생성으로 객채 생성 비용 절약하였습니다.

kimhyochan committed:
[feat] 중복없는 무작위 숫자 6개를 생성하는 메소드 작성

kimhyochan committed:
[feat] 생성된 모든 LottoNumbers 출력

kimhyochan committed:
[feat] LottoNumbers의 목록을 가지고 있는 일급 컬렉션 생성

kimhyochan committed:
[feat] 당첨 번호와 보너스 볼을 관리하는 클래스 생성

int형의 번호를 알맞은 Enum 타입으로 파싱하는 메서드 생성

kimhyochan committed:
[feat] 당첨 번호 중복 예외 처리

kimhyochan committed:
[feat] 구매한 모든 로또 숫자들의 유효성을 검증하는 메서드 추가

kimhyochan committed:
[fix] WinningNumbers의 생성 중 중복 체크 중 추가되는 버그 수정

kimhyochan committed:
[feat] 지난 주 당첨번호와 보너스 볼을 입력받는 기능

kimhyochan committed:
[feat] 당첨 결과를 담는 enum 생성

kimhyochan committed:
[feat] 어플리케이션 메인 로직 생성

Commit:
Merge pull request #1 from hyochan-Kim/lotto

Lotto 풀리퀘 드립니다!

Commit:
[refact] test 폴더 루트 리팩토링

Commit:
[refact] Map 캐시 전략을 이용한 싱글톤 LottoNumber 생성

기존의 클래스를 한번만 생성하기 위해 사용했던
Enum LottoNumberGroup에서 LottoNumber 클래스로 변경하며
Map을 사용한 팩토리 메서드를 활용하기로 하였습니다.

같은 기능을 하지만 Enum 타입은 보일러 플레이트 코드가 너무
많아져 변경하게되었다.

Commit:
[refact] LottoNumber 타입 리팩토링

LottoNumber의 타입을 변경하면서 발생하는 에러를 리팩토링하였다.

Commit:
[refact] LottoNumbers, WinningNumbers 상속 리팩토링

LottoNumbers와 WinningNumbers의 중복 사항들을 LottoRound
클래스로 옮겨 중복을 제거하고 상속받게 리팩토링하였다.

Commit:
[refact] WinningNumbersTest 타입 리팩토링

LottoNumber의 타입을 변경하면서 발생하는 에러를 리팩토링하였다.

Commit:
[feat] GameResult 계산 기능

interface를 추출하여 알맞은 GameResult를 추출하는 기능을 추가하였다.

Commit:
[feat] 패키지 리팩토링

가독성을 위해 공통적인 로직을 가지고 있는 디렉토리 위치를 변경하였다.

Commit:
[feat] 테스트 패키지 리팩토링

가독성을 위해 공통적인 로직을 가지고 있는 디렉토리 위치를 변경하였다.

Commit:
[feat] GameResult 출력 기능 추가

GameResult를 출력하기 위해 GameResult 판다 논리를 담은
람다식 매개변수를 맞은 로또 넘버의 수로 바꿔
GameResult를 출력하기 용이하게 변경하였다.

Commit:
[feat] 수익률 계산

Commit:
[feat] 수익률 출력

Commit:
[feat] 당첨 번호와 로또 번호 중 맞는 수 카운트

WinnginNumbers에 당첨번호와 로또 번호 중 맞는 수를
카운팅하는 기능을 추가했다.

Commit:
[feat] 보너스 번호가 맞는지 확인하는 기능

Commit:
[fix] int 범위를 넘어가는 이익의 수익률 계산

int범위를 넘어가는 이익이 들어올 때 수익률 계산 에러 해결
double형으로 교체하였다.

Commit:
[refact] 랜덤넘버생성기 테스트 반복 추가

Commit:
[refact] main 메서드 분리 리팩토링

Commit:
[refact] inputView 발생 예외 커스텀

예외를 catch하여 알맞은 동작을 구동시키기 위해
커스텀 Exception을 생성하였다.

Commit:
[fix] 당첨 통계 금액 부동소수점 표기 수정

Commit:
[refact] AllLottoNnumbers에 GameResult 리스트 책임 부여

Application에서 가지고 있던 책임을 AllLottoNumbers에게 위임

Commit:
[fix] LottoNumber 정렬

Commit:
[refact] 자바 컨벤션

Commit:
[refact] 테스트 코드 리팩토링

Commit:
[refact] 테스트 코드 리팩토링

Commit:
[refact] LottoRound 자료형 접근지정자 변경

Commit:
[refact] AllLottoNumbers Null 체크 로직 변경

Commit:
[feat] 구매 금액 이하의 값 입력 시 예외 처리

jihan805 commented on src/main/java/lotto/domain/number/PurchaseNumber.java:
```diff
@@ -0,0 +1,34 @@
+import lotto.domain.exception.PurchaseMoneyLackException;
+
+public class PurchaseNumber {
+    private static final int TICKET_PRICE = 1000;
```
1. 돈을 나타내는 숫자의 경우 천 단위를 구분하기 위해 "1_000"와 같이 언더바를 통해 구분!
코드 전체적으로 적용해보면 어떨까요?

2. PurchaseNumber 보다는 구입 금액과 더 관련이 있는 클래스 인 것 같아요. 
PurchaseNumber와 Yield 를 나누기 보다, 입력하는 돈, 수익률 계산 등 돈과 관련한 모든 로직을 Money와 같은 객체를 추가해 구현해 보는 것은 어떨까?요

jihan805 commented on src/main/java/lotto/ConsoleLottoApplication.java:
```diff
@@ -0,0 +1,61 @@
+        return purchaseNumber;
+    }
+
+    private static AllLottoNumbers createAllLottoNumbers(PurchaseNumber purchaseNumber) {
```
로또 번호 생성을 위한 역할이 Application의 역할로 적절할까요?
로또 번호 생성을 위한 클래스 추출을 통해 "로또 번호 생성" 이라는 역할을 위임해보면 어떨까요?

jihan805 commented on src/main/java/lotto/domain/number/LottoRound.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+import java.util.Objects;
+
+public class LottoRound {
```
코드 재사용에는 상속과 조합(composition) 두가지 방식이 있어요. 
상속과 조합의 장단점에 대해 찾아보면 어떨까요?

jihan805 commented on src/main/java/lotto/domain/number/LottoRound.java:
```diff
@@ -0,0 +1,31 @@
+    public static final int LOTTO_NUMBER_AMOUNT = 6;
+    private static final String LOTTO_NUMBER_AMOUNT_EXCEPTION_MESSAGE = "로또 번호의 갯수가 6개가 아닙니다.";
+
+    List<LottoNumber> lottoNumbers;
```
List 자료구조가 아닌 Set 자료구조를 활용하는 것은 어떨까요?
Set을 활용했을 때 얻을 수 있는 이점은 무엇이 있을까요?

jihan805 commented on src/main/java/lotto/domain/number/AllLottoNumbers.java:
```diff
@@ -0,0 +1,41 @@
+        }
+    }
+
+    public List<GameResult> calcurateGameResult(WinningNumbers winningNumbers) {
```
calcurateGameResult 오타 !

jihan805 commented on src/main/java/lotto/ConsoleLottoApplication.java:
```diff
@@ -0,0 +1,61 @@
+
+    private static GameResults createGameResults(AllLottoNumbers allLottoNumbers, WinningNumbers winningNumbers) {
+        List<GameResult> gameResults = allLottoNumbers.calcurateGameResult(winningNumbers);
+        return new GameResults(gameResults);
```
바로 게임 결과를 구하기보다 LottoGame/ LottoPlayer등의 클래스를 통해 게임 결과를 구해보면 어떨까요?

jihan805 commented on src/main/java/lotto/domain/number/WinningNumbers.java:
```diff
@@ -0,0 +1,39 @@
+                .count();
+    }
+
+    public boolean isCorrectBonusNumber(LottoNumbers lottoNumb2ers) {
```
1. lottoNumb2ers -> lottoNumbers
2. is~ 를 사용하는 boolean 네이밍에 대해 찾아보면 어떨까요? (is + 명사/형용사/ing)

jihan805 commented on src/test/java/lotto/domain/number/PurchaseNumbersTest.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Test
+    @SuppressWarnings("NonAsciiCharacters")
+    void calculate_테스트() {
```
2000원으로 2개를 구매할 수 있다면 총 구매 수가 2인 것을 검증하는 것이 좋지 않을까요?

jihan805 commented on src/test/java/lotto/domain/number/AllLottoNumbersTest.java:
```diff
@@ -0,0 +1,24 @@
+        assertThatThrownBy(() -> {
+            AllLottoNumbers allLottoNumbers = new AllLottoNumbers(Collections.emptyList());
+        }).isInstanceOf(IllegalArgumentException.class);
+    }
```
현재 해당 클래스에는 구입로또와 우승로또의 결과를 구하는 calcurateGameResult() 메서드가 있어요.
해당 테스트가 보이지 않는 것 같아요!

jihan805 reviewed (changes requested):
안녕하세요~! 전체적으로 구현 잘하셨어요 💯 
몇가지 피드백 남겼으니 확인해주세요 :)

Commit:
[refact] 천 단위 구별을 위한 "_" 추가

Commit:
[refact] Yield, PurchaseNumber를 Money로 리팩토링

수익률을 의미하는 Yield, 로또 라운드 수를 의미하는 PurchaseNumber가
Money라는 이름의 금액으로 추상화될 수 있다고 판단하여
하나의 클래스에서 책임지도록 변경하였다.

Commit:
[refact] AllLottoNumbers, LottoNumbers, LottoNumber 클래스명 변경

구매한 모든 로또 라운드의 수를 가지고 있는 AllLottoNumbers와
라운드 하나의 로또 숫자를 가지고 있는 LottoNumbers와
1부터 45의 로또를 구성하는 숫자를 가지고 있는 LottoNumber의
작명이 혼동을 일으킨다고 판단하여 각 LottoRounds, LottoRound로
변경하였다.

또한 LottoRound를 LottoRoundFormat으로 변경하여 한 라운드의
규격을 나타내려는 의도를 보이도록 하였다.

Commit:
[refact] 로또 번호 생성 클래스 추출

LottoRoundsGenerator 클래스를 통해 로또 번호 생성하는 클래스를
추출하여 책임 부여하였다.

Commit:
[refact] LottoRound Set 자료구조 변경

중복을 허용하지 않는 점에서 List 보다 Set이 중복에 대한
예외를 처리하기 용이하다고 판단하여 Set으로 변경하였다.

Commit:
[refact] boolean 메서드 네이밍 변경

보너스 번호를 가지고 있는지 확인하는 메서드에서 is에서
가지고 있다를 의미하는 has로 메서드 명을 변경하였습니다.

Commit:
[refact] 게임 결과 도출 테스트 추가

Commit:
[refact] LottoGame 클래스 추가

Main은 입출력과 게임 실행을 이어주는 기능만을 담당해야한다고
생각하여 게임의 흐름을 제어하는 LottoGame을 생성하였다.

Commit:
[refact] LottoRoundFormat 삭제

WinnningNumbers, LottoRound 에서 동일하게 사용하는 여섯개의
숫자를 나타내는 자료형을 둘다 다루는 것에서 LottoRoundFormat을
상속받는 형태로 구현하였으나,

상속보다는 유연한 객체 조립으로 해당 기능을 리팩토링하였다.

Commit:
[refact] 컨벤션 리팩토링

kingbbode commented on src/main/java/lotto/domain/result/Money.java:
```diff
@@ -0,0 +1,48 @@
+        return (int) money / TICKET_PRICE;
+    }
+
+    public double calculateYield(GameResults gameResults) {
+        double purchaseMoney = money;
+        double benefit = gameResults.calculateBenefit();
+        return (benefit / purchaseMoney) * MULTIPLE_PERCENTAGE;
+    }
```
의존 관계를 반대로 해보는 것은 어떨까요?

`Money` 가 `GameResults` 를 의존하기에는 `GameResults` 의 정보가 더 많은 것 같아요 :)

`Money` 의 역할로 생각을 해보아도 이 계산은 `Money`의 일이 아니지 않을까 생각합니다.

kingbbode reviewed (approved):
피드백이 잘 반영되었네요!
다음 단계도 잘 부탁드려요.
코멘트는 참고 부탁드려요 :)
